x/stakeibc/keeper/icacallbacks_undelegate.go

```diff
@@ -204,31 +202,42 @@ func (k Keeper) UpdateHostZoneUnbondingsAfterUndelegation(
 	ctx sdk.Context,
 	chainId string,
 	epochUnbondingRecordIds []uint64,
-	totalStTokensToBurn sdkmath.Int,
 	totalNativeTokensUnbonded sdkmath.Int,
 	unbondingTime uint64,
-) error {
+) (totalStTokensToBurn sdkmath.Int, err error) {
+	// As we process the accounting changes, keep track of the stTokens that should be burned later
+	totalStTokensToBurn = sdkmath.ZeroInt()
+
 	// Loop each epoch unbonding record starting from the earliest
 	for _, epochNumber := range epochUnbondingRecordIds {
 		hostZoneUnbonding, found := k.RecordsKeeper.GetHostZoneUnbondingByChainId(ctx, epochNumber, chainId)
 		if !found {
-			return errorsmod.Wrapf(recordstypes.ErrHostUnbondingRecordNotFound,
+			return totalStTokensToBurn, errorsmod.Wrapf(recordstypes.ErrHostUnbondingRecordNotFound,
 				"host zone unbonding not found for epoch %d and %s", epochNumber, chainId)
 		}
 
-		// Determine the amount to decrement from each record, flooring at 0
-		stTokensToBurn := sdkmath.MinInt(hostZoneUnbonding.StTokensToBurn, totalStTokensToBurn)
+		// Determine the native amount to decrement from the record, capping at the amount in the record
+		// Also decrement the total for the next loop
 		nativeTokensUnbonded := sdkmath.MinInt(hostZoneUnbonding.NativeTokensToUnbond, totalNativeTokensUnbonded)
-
-		// Decrement the records
-		hostZoneUnbonding.StTokensToBurn = hostZoneUnbonding.StTokensToBurn.Sub(stTokensToBurn)
 		hostZoneUnbonding.NativeTokensToUnbond = hostZoneUnbonding.NativeTokensToUnbond.Sub(nativeTokensUnbonded)
-
-		// Update the totals for the next loop
-		totalStTokensToBurn = totalStTokensToBurn.Sub(stTokensToBurn)
 		totalNativeTokensUnbonded = totalNativeTokensUnbonded.Sub(nativeTokensUnbonded)
 
-		// If there are no more tokens to burn after this batch, iterate the record to the next status
+		// Calculate the relative stToken portion using the implied RR from the record
+		// If the native amount has already been decremented to 0, just use the full stToken remainder
+		// from the record to prevent any precision error
+		var stTokensToBurn sdkmath.Int
+		if hostZoneUnbonding.NativeTokensToUnbond.IsZero() {
+			stTokensToBurn = hostZoneUnbonding.StTokensToBurn
+		} else {
+			impliedRedemptionRate := sdk.NewDecFromInt(hostZoneUnbonding.NativeTokenAmount).Quo(sdk.NewDecFromInt(hostZoneUnbonding.StTokenAmount))
+			stTokensToBurn = sdk.NewDecFromInt(nativeTokensUnbonded).Quo(impliedRedemptionRate).TruncateInt()
```

Comment:
Whereas normally I worry about Int Truncates leaving us with unpredictable roundings, but beautifully the other case here should make it not a problem -- there will be remaining stTokens only in the case now when there is remaining native, and when this record gets finished off and native is zero, the other case sets the stTokensToBurn to exactly the stTokensToBurn so any previous rounding issues wouldn't matter anyway. Nice